packages/contracts/src/SequencerSet.sol

```diff
@@ -0,0 +1,172 @@
+        require(OperatorStake[_chainId][msg.sender] >= minimumStakeThreshold);
+
+        uint256 sequencerStateLength = SequencerStates[_chainId].length;
+        SequencerState memory latestSequencerState;
+
+        if (sequencerStateLength == 0) {
+            latestSequencerState = SequencerState({
+                blockNumber: block.number - 1,
+                aggregatedPublicKey: initAggregatedPublicKey,
+                aggregatedPublicKeyHash: initAggregatedPublicKeyHashed
+            });
+        } else {
+            latestSequencerState = SequencerStates[_chainId][
+                sequencerStateLength - 1
+            ];
+        }
+        uint256[6] memory _aggregatedPublicKey = BLS.addJac(
+            [publicKey[0], publicKey[1], publicKey[2], publicKey[3], 1, 0],
+            [
+                latestSequencerState.aggregatedPublicKey[0],
+                latestSequencerState.aggregatedPublicKey[1],
+                latestSequencerState.aggregatedPublicKey[2],
+                latestSequencerState.aggregatedPublicKey[3],
+                1,
+                0
+            ]
+        );
+
+        bytes32 publicKeyHashed = keccak256(
+            abi.encode(publicKey[0], publicKey[1], publicKey[2], publicKey[3])
+        );
+
+        require(PubKeyHashToOperator[_chainId][publicKeyHashed] == address(0));
+        require(OperatorToPubKeyHash[_chainId][msg.sender] == 0);
+
+        OperatorToPubKeyHash[_chainId][msg.sender] = publicKeyHashed;
+        PubKeyHashToOperator[_chainId][publicKeyHashed] = msg.sender;
+
+        bytes32 apkHash = keccak256(
+            abi.encodePacked(
+                _aggregatedPublicKey[0],
+                _aggregatedPublicKey[1],
+                _aggregatedPublicKey[2],
+                _aggregatedPublicKey[3]
+            )
+        );
+
+        if (latestSequencerState.blockNumber == block.number) {
+            // update this array
+            SequencerStates[_chainId][
+                sequencerStateLength - 1
+            ] = SequencerState({
+                blockNumber: block.number,
+                aggregatedPublicKey: [
+                    _aggregatedPublicKey[0],
+                    _aggregatedPublicKey[1],
+                    _aggregatedPublicKey[2],
+                    _aggregatedPublicKey[3]
+                ],
+                aggregatedPublicKeyHash: apkHash
+            });
+        } else {
+            // update this array
+            SequencerStates[_chainId].push(
+                SequencerState({
+                    blockNumber: block.number,
+                    aggregatedPublicKey: [
+                        _aggregatedPublicKey[0],
+                        _aggregatedPublicKey[1],
+                        _aggregatedPublicKey[2],
+                        _aggregatedPublicKey[3]
+                    ],
+                    aggregatedPublicKeyHash: apkHash
+                })
+            );
+        }
```

Comment:
Gas-costly because we expand the array so write to new storage slots, but it's Gnosis, it's cheaaaap.